sources/curriculum/0_before_you_start/workflows-and-pipelines-detailed/README.md

```diff
@@ -0,0 +1,194 @@
+
+If the code base is not structured and named well, you might be struggling
+to remember the details of each step once you have built a few models. 
+What features did you use for each?
+What training and testing split?  What hyperparameters?
+
+Your code might be getting messy too. Did you overwrite the code for
+the previous model? Maybe you copied, pasted, and edited code from an
+earlier model. Can you still read what's there? It can quickly become
+a hodgepodge that requires heroics to decipher.
+
+In this session, we will introduce a workflow that can avoid (or at least reduce) these problems.
+
+# Data pipelines
+
+It is helpful to structure the data into multiple layers.
+
+##### Raw
+The data we receive from the partners and external sources is the raw data. 
+Raw data is immutable. Quoting from the popular workflow package [Data Science Cookiecutter](https://drivendata.github.io/cookiecutter-data-science/#data-is-immutable):
+
+Don't ever edit your raw data, especially not manually, and especially not in Excel. 
+Don't overwrite your raw data. 
+Don't save multiple versions of the raw data. 
+Treat the data (and its format) as immutable. 
+The code you write should move the raw data through a pipeline to your final analysis. 
+You shouldn't have to run all of the steps every time you want to make a new figure (see Analysis is a DAG), but anyone should be able to reproduce the final products with only the code in src and the data in data/raw.
+
+
+##### Intermediate
+If the raw data is messy, it is advisable to create an intermediate layer that consists of tidy copies of the raw data.
+Typical situations where this is useful are:
+- Data is received in multiple different file types 
+- Data fields are not typed (e.g. csv files, excel) or poorly typed (dates as strings, inconsistent date formats)
+- Column names are unclear, have spaces, special characters, or there are no column names 
+
+The transformations from raw to intermediate should be limited to fix the issues mentioned above. 
+We should not combine different data sets or create calculated fields.
+This is reserved for the next layer.
+
+Typical file formats for the intermediate layer are a data base (e.g. `postgres`) or `parquet` files.
+
+##### Processed
+To perform the modelling work, the input data needs to be combined and enriched, for example by creating features. 
+The data sets that are created in this process are stored in the processed layer.
+Sometimes it can be useful to split this layer out into a domain data model, a feature layer and a master layer but
+the exact layering will depend on the project context.
+
+##### Models
+The processed data is used to train predictive models, explanatory models, recommender engines and optimisation algorithms.
+The trained models are stored in the model layer. 
+In contrast to the previous layers, models are usually stored in `pickle` because they are not in tabular format.
+
+##### Reporting
```

Comment:
In DSaPP we use `results` instead of `reporting`. The rationale about this is that we would do, postmodeling, modeling selection, bias analysis, etc.

Reply:
In my view the reporting layer contains the model evaluation metrics but is wider since it contains any output you may want to display in the front end. 
I have seen teams that split reporting into model output and reporting. That works, too.

Reply:
Ok, please adjust

Reply:
done.